File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode BuildTree(int[] preorder, int[] inorder) {
        Dictionary<int,int> inorderIdx = new Dictionary<int,int>();
		
        for(int i=0;i<inorder.Length;i++)
            inorderIdx[inorder[i]] = i;
        
		int preIdx = 0;
        return Build(0,preorder.Length-1);
		
        TreeNode Build(int inStartIdx, int inLastIdx)
        {
            if(inStartIdx>inLastIdx) return null;
            
            TreeNode root = new TreeNode(preorder[preIdx]);
            int rootIdx = inorderIdx[preorder[preIdx++]];
            root.left = Build(inStartIdx,rootIdx-1);
            root.right = Build(rootIdx+1,inLastIdx);
            return root;
        }
    }
}

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
map<int, int> inMap;
    int pos = 0;
    TreeNode* helper(int start, int end, vector<int>& preorder) {
        
        if(start > end) return nullptr;
        int val = preorder[pos];
        TreeNode *node = new TreeNode(val);
        
        int inPos = inMap[preorder[pos]];
        pos++;
        

        node->left = helper(start, inPos - 1, preorder);

        node->right = helper(inPos + 1, end, preorder);

        return node;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        for(int i = 0; i < inorder.size(); i++) {
            inMap[inorder[i]] = i;
        };
        
        int l = 0;
        int r = preorder.size() - 1;

        return helper(l, r, preorder);
    }
};